atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CqlExecutor.java

```diff
@@ -142,4 +152,75 @@ private CellWithTimestamp getCell(byte[] key, CqlRow cqlRow) {
 
         return new CellWithTimestamp.Builder().cell(Cell.create(key, columnName)).timestamp(timestampLong).build();
     }
+
+    private final class CqlQuery {
+        private final String queryFormat;
+        private final Arg<?>[] queryArgs;
+
+        CqlQuery(String queryFormat, Arg<?>... args) {
+            this.queryFormat = queryFormat;
+            this.queryArgs = args;
+        }
+
+        public List<CellWithTimestamp> executeAndGetCells(byte[] row) {
+            CqlResult cqlResult = KvsProfilingLogger.maybeLog(
+                    () -> queryExecutor.execute(row, toString()),
+                    this::logSlowResult,
+                    this::logResultSize);
+            return getCells(row, cqlResult);
+        }
+
+        private void logSlowResult(KvsProfilingLogger.LoggingFunction log, Stopwatch timer) {
+            Object[] allArgs = new Object[queryArgs.length + 3];
+            allArgs[0] = SafeArg.of("queryFormat", queryFormat);
+            allArgs[1] = UnsafeArg.of("fullQuery", toString());
+            allArgs[2] = LoggingArgs.durationMillis(timer);
+            System.arraycopy(queryArgs, 0, allArgs, 3, queryArgs.length);
+
+            log.log("A CQL query was slow: queryFormat = [{}], fullQuery = [{}], durationMillis = {}", allArgs);
```

Comment:
so `fullQuery` actually contains the whole query (as an `UnsafeArg`). large internal product will get that substituted in to the braces. they will miss all the parameters, but they don't need them because they already have the full query